### > ### **Assignment 2 - Numpy Array Operations** 


# Some NumPy Functions


### NumPy seems Fun, NGL!!!

Write a short introduction about Numpy and list the chosen functions. 

- function 1 - numpy.prod()
- function 2 - numpy.sort()
- function 3 - numpy.median()
- function 4 - numpy.char.add()
- function 5 - numpy.swapaxis()

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [104]:
!pip install jovian --upgrade -q

In [105]:
import jovian

In [106]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [107]:
import numpy as np

In [108]:
# List of functions explained 
function1 = np.prod  # (change this)
function2 = np.sort
function3 = np.median
function4 = np.mean
function5 = np.histogram

## Function 1 - np.prod

So basically this numPy function will return the product of array element of a given axis

Ex: np.prod(a, axis= ?, datatype), So here a is an array/Matrix and then in axis we have to give some value and it will return product of that perticular elements of axis

In [109]:
# Example 1 - working
arr1 = [[1, 2], 
        [3, 4.]]

arr2 = [[5, 6, 7], 
        [8, 9, 10]]

prod1 = np.prod(arr1,1)
prod2 = np.prod(arr2,1)
print(prod1)
print(prod2)

[ 2. 12.]
[210 720]


#### Explanation about example

And we can say for rows we can use 0 and for columns we can use 1 as an axis, right? Like in most common cases

So above as you can see if we use:
                                  
np.prod(arr1,1), so what it'll do --> It will do product of 1*2=2 and 3*4 = 12

np.prod(arr2,1), so what it'll do --> It will do product of 5*6*7=210 and 8*9*10 = 720 and it will then retrun the output

In [110]:
# Example 2 - working
arr3 = np.array([[1,2,3],
                [4,5,6],
                [7,8,9],
                [3234,34323,23]])
prod3 = np.prod(arr3,0)
print(prod3)

prod4 = np.prod(arr3,1)
print(prod4)

[  90552 2745840    3726]
[          6         120         504 -1741953910]


So above as you can see if we use:
                                  
np.prod(arr3,0), so what it'll do --> It will do product of elements of rows because axis is 0 so 3234*7*4*1... and so on

np.prod(arr3,1), so what it'll do --> It will do product of elements of columns because axis is 1 so 3*2*1... and so on

In [111]:
# Example 3 - breaking (to illustrate when it breaks)
arr4 = np.random.randn(3,3,3)
print(arr4)

prod5 = np.prod(arr4,3)
#So Here I have used random.randn to create random 3D array

[[[ 0.63233097 -1.65748041  0.31636799]
  [-1.02229794 -0.49194315  2.21129205]
  [-0.90971228 -1.61045483  1.4405375 ]]

 [[ 0.48148762 -0.80874648 -1.12859056]
  [-0.77725391  0.08563175 -1.00245148]
  [-0.81968295  0.40568413 -1.40338923]]

 [[-0.24986559  0.55610976 -0.57832143]
  [ 0.73340785  2.81985689 -0.76645518]
  [-1.61285746  0.77787142  1.11916806]]]


AxisError: axis 3 is out of bounds for array of dimension 3

Explanation about example (why it breaks and how to fix it)

So here dimension of matrix is 3 and in axis we have entered 3 and in NumPy indexing starts from 0 so what we can do is we can set axis vlaue between -3 to 2

Some closing comments about when to use this function.

In [112]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

## Function 2 - numpy.sort

It returns sorted copy of an array and we can also sort some perticular axis by giving value of axis in parameters in sort function. 

numpy.sort(a,axis=?)

So here a = numpy Array
 and axis = Accordingly

In [113]:
arr3

array([[    1,     2,     3],
       [    4,     5,     6],
       [    7,     8,     9],
       [ 3234, 34323,    23]])

In [114]:
arr3.sort()
print(arr3)

[[    1     2     3]
 [    4     5     6]
 [    7     8     9]
 [   23  3234 34323]]


So what sort function will do is it will sort the elements inside arr3 matrix and it will return sorted Matrix, since we have not given any axis in parameters it will automatically be -1 which will be last

In [115]:
arrSortEx = np.array([[3,4,35],
                     [33,4342,13],
                     [23,3,424],
                     [53,3124,131],
                     [54232,23,23]])
print(arrSortEx)
np.shape(arrSortEx)

[[    3     4    35]
 [   33  4342    13]
 [   23     3   424]
 [   53  3124   131]
 [54232    23    23]]


(5, 3)

In [116]:
np.sort(arrSortEx,0)

array([[    3,     3,    13],
       [   23,     4,    23],
       [   33,    23,    35],
       [   53,  3124,   131],
       [54232,  4342,   424]])

In [117]:
np.sort(arrSortEx,1)

array([[    3,     4,    35],
       [   13,    33,  4342],
       [    3,    23,   424],
       [   53,   131,  3124],
       [   23,    23, 54232]])

So here in np.sort(arrSortEx,1) it will sort elements column wise and in np.sort(arrSortEx,0) it will sort element row wise and if given axis is none then it will sort by default (-1)

In [118]:
# Example 3 - breaking (to illustrate when it breaks)
np.sort(arrSortEx,2)

AxisError: axis 2 is out of bounds for array of dimension 2

Explanation about example (why it breaks and how to fix it)

So it breakes because we have given value of axis = 2, but it's not valid because of dimensions of array, the dimension here is 2 and we have given value of axis = 2, so It gives error that axis 2 is out of bound and to fix this we can simply change value of axis in between -2 to 1

Some closing comments about when to use this function.

In [119]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

## Function 3 - numpy.mean

Well in maths mean means sum of elements/number of elements so we can say sum(n)/n but in matrixes to find mean NumPy gives us mean function, using this we can find mean simply.

mean(a, axis=?)

and above function will return mean of Matrix

In [120]:
mean1 = np.array([[4,34,2],
                 [324,4242,32],
                 [4322,3131,13],
                 ])

In [121]:
mean1

array([[   4,   34,    2],
       [ 324, 4242,   32],
       [4322, 3131,   13]])

In [122]:
np.mean(mean1,1)

array([  13.33333333, 1532.66666667, 2488.66666667])

In [123]:
np.mean(mean1)

1344.888888888889

So what basically above function does is, it gives us mean of all columns elements because axis = 1

In [124]:
# Example 2 - working
np.mean(mean1,0)

array([1550.        , 2469.        ,   15.66666667])

So what this does is, it gives us mean of all rows and returns the answer

In [125]:
# Example 3 - breaking (to illustrate when it breaks)
np.mean(mean1,2)

AxisError: axis 2 is out of bounds for array of dimension 2

Explanation about example (why it breaks and how to fix it)

Here the Array is 2-D and we have taken axis 2 which is out of bound so it gives us error, we can fix it by changing axis between -2 to 1

Some closing comments about when to use this function.

In [126]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

## Function 4 - numpy.char.add

Add some explanations

In [127]:
# Example 1 - working
arr1 = ([["Serie A: "],
        ["La Liga: "],
        ["PL: "],
        ["BundesLiga: "]])

arr2 = ([["CR7 is king of SA ofc"],
        ["Messi is the GOD of Football"],
        ["KDB is face of PL I said what I said"],
        ["RL9 is face of Polish league and no one comes close except HAAALLAAAND"]])

np.char.add(arr1,arr2)

array([['Serie A: CR7 is king of SA ofc'],
       ['La Liga: Messi is the GOD of Football'],
       ['PL: KDB is face of PL I said what I said'],
       ['BundesLiga: RL9 is face of Polish league and no one comes close except HAAALLAAAND']],
      dtype='<U82')

It just adds two arrays and then displays the added array of both :O

In [128]:
# Example 2 - working
arr1 = np.char.add(arr1,"  ")
np.char.add(arr1,arr2)

array([['Serie A:   CR7 is king of SA ofc'],
       ['La Liga:   Messi is the GOD of Football'],
       ['PL:   KDB is face of PL I said what I said'],
       ['BundesLiga:   RL9 is face of Polish league and no one comes close except HAAALLAAAND']],
      dtype='<U84')

In this I have added some space and added both Matrixes

In [129]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = ([[32,32,113],
        [32,2,332],
        [54,3,32],
        [23,231,1]])

arr2 = ([["Serie A: "],
        ["La Liga: "],
        ["PL: "],
        ["BundesLiga: "]])

np.char.add(arr1,arr2)

TypeError: string operation on non-string array

We can't add both array because one of them has int types and given function only works when type is String so what we can do to fix it is, we can add numbers as characters and it will be fixed

Well you can use this function when you're working with large datasets and you want to add some counties using some indexes so you can use this

In [130]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

## Function 5 - numpy.swapaxis

We can interchange axis of an array

np.swapaxes(a,axis1,axis2) so axis 1 will be interchanged with axis 2

In [131]:
# Example 1 - working
arr1 = np.arange(27).reshape(3,3,3)
print("Given array is: \n")
print(arr1,"\n")
print("After swap of axis the array looks like: \n")
print(np.swapaxes(arr1,2,0))

Given array is: 

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]] 

After swap of axis the array looks like: 

[[[ 0  9 18]
  [ 3 12 21]
  [ 6 15 24]]

 [[ 1 10 19]
  [ 4 13 22]
  [ 7 16 25]]

 [[ 2 11 20]
  [ 5 14 23]
  [ 8 17 26]]]


We have interchanged columns

In [132]:
# Example 2 - working
b = np.arange(8).reshape(2,2,2)
print("Given array is \n\n",b,"\n")
print("After swap array is \n",np.swapaxes(b,1,2),"\n")

Given array is 

 [[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]] 

After swap array is 
 [[[0 2]
  [1 3]]

 [[4 6]
  [5 7]]] 



We changed columns of given 3-D Matrix

In [133]:
# Example 3 - breaking (to illustrate when it breaks)
c = np.arange(8).reshape(2,2,2)
print("Given array is \n\n",c,"\n\n")
print("After swap array is: ",np.swapaxes(c,3,2))

Given array is 

 [[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]] 




AxisError: axis1: axis 3 is out of bounds for array of dimension 3

Here it gives us error because Array is 3-D and we are trying to use 3rd axis but it's not possible since dimension is 3, so we can see the shape and set axis accordingly that will fix the issue

Some closing comments about when to use this function.

In [134]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

In [135]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

## Conclusion

Summarize what was covered in this notebook, and where to go next

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* ...

In [136]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations


'https://jovian.ai/mvbosamiya20/numpy-array-operations'

In [137]:
jovian.submit(assignment="zero-to-pandas-a2")

<IPython.core.display.Javascript object>

[jovian] Updating notebook "mvbosamiya20/numpy-array-operations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/mvbosamiya20/numpy-array-operations
[jovian] Submitting assignment..
[jovian] Verify your submission at https://jovian.ai/learn/data-analysis-with-python-zero-to-pandas/assignment/assignment-2-numpy-array-operations
